### PR TITLE
Update MacOS installation directory

### PR DESCRIPTION
**Pull request by MDrakos**
https://www.pivotaltracker.com/story/show/177749041

### CODE REVIEW

File: cmd/state-installer/internal/installer/installation_darwin.go
Comment (mdrohmann) [inline cmd/state-installer/internal/installer/installation_darwin.go:38]: So, the tray app is the only one, we need to symlink now?  That is good...

File: cmd/state-installer/internal/installer/installation_darwin.go
Comment (mdrohmann) [inline cmd/state-installer/internal/installer/installation_darwin.go:22]: I think in an update, this would error out, because the directory already exists.  We somehow need to make this function able to run twice without error. The easiest way is probably to just remove the entire target directory "ActiveState Desktop.app" at the beginning.

Of course it would be nice to also have a rollback function for this, but that is probably better done in a follow-up story, as my base PRs haven't landed yet.  And it will be easier once that has happened.

**Review by mdrohmann — CHANGES_REQUESTED** [review]
I realized that the function needs to be idempotent.  So, we should clean up the app dir before we write files in there.

File: cmd/state-installer/internal/installer/installation_darwin.go
Comment (MDrakos) [reply cmd/state-installer/internal/installer/installation_darwin.go:46]: Yes, `appinfo` takes care of that so we only need to like the `state-tray` binary

File: cmd/state-installer/internal/installer/installation_darwin.go
Comment (MDrakos) [reply cmd/state-installer/internal/installer/installation_darwin.go:30]: Actually, this function will just return `nil` if the directory already exists. But I think it's better to just remove the directory and all of its contents rather than overwriting everything so I've applied this suggestion.

Is there a follow-up story for rollback functionality in our update/install mechanism already? If not I can create one.

**Review by mdrohmann — DISMISSED** [review]

File: cmd/state-installer/internal/installer/installation_darwin.go
Comment (MDrakos) [reply cmd/state-installer/internal/installer/installation_darwin.go:30]: Ticket is here: https://www.pivotaltracker.com/story/show/177873388

File: cmd/state-installer/internal/installer/installation_darwin.go
Comment (Naatan) [inline cmd/state-installer/internal/installer/installation_darwin.go:47]: These paths need variables so this is more readable. There's way too much happening in these args.

File: cmd/state-installer/internal/installer/installation_darwin.go
Comment (Naatan) [inline cmd/state-installer/internal/installer/installation_darwin.go:56]: This should use `TargetExists`. Symlinks aren't technically "files" (ie. this won't be true if the `filename` is a symlink).

File: cmd/state-installer/internal/installer/installation_darwin.go
Comment (mdrohmann) [reply cmd/state-installer/internal/installer/installation_darwin.go:27]: Mike and I discussed that a little.  I initially liked the idea to *only* ship the executables in the archive.  On the other hand, we already noticed that we'll need a customary rollback mechanism for this.

So, thinking this through again, it is probably better if we stick with one system as @Naatan suggested.

Comment (Naatan) [reply cmd/state-installer/internal/installer/installation_darwin.go:27]: Ultimately I'll defer to you on this @mdrohmann since you're owning the update process and I can see this significantly delaying the PR. Could potentially be a later iteration also. But yeah I do think it would be better if this wasn't embedded in the binary.

File: cmd/state-installer/internal/installer/installation_darwin.go
Comment (mdrohmann) [reply cmd/state-installer/internal/installer/installation_darwin.go:47]: The latest `appinfo` implementation on `master` should have an option to set a custom base directory, so you can say:

```
fromTray := appinfo.TrayApp(installDir)
toTray := appinfo.TrayApp(appDir)
createSymink(fromTray.Exec(), toTray.Exec())
```

File: scripts/update-generator/main.go
Comment (mdrohmann) [inline scripts/update-generator/main.go:166]: This works, but it is rather implicit logic: files are binaries => `./bin` and directories are system files that go in the `./system` directory.  Can we maybe provide the system files as a separate argument?  Then we would need to introduce a flag for this, something like `-systemApplicationDir`...

**Review by mdrohmann — CHANGES_REQUESTED** [review]

**Comment by mdrohmann** [comment]
I realized that for these integration tests to work: https://github.com/ActiveState/cli/runs/2423391294#step:14:88, you'll probably need to add the app dir to the `generate-test-update` script too.  You should be able to test them locally on your own machine, BTW...

**Review by mdrohmann — APPROVED** [review]

**Comment by mdrohmann** [comment]
Well, I noticed that the Critical Integration Tests on MacOS are still failing. I know they are a little bit difficult to fix, but I found this workflow useful:

- Change the "retainDirectories" flag to `true` (eg., [here](https://github.com/ActiveState/cli/blob/master/test/integration/update_int_test.go#L281-L281)
- Run the test locally
- Read the installation log file that appears in the test output

As it only fails on MacOS now, it probably fails because of some issue with the added functionality.

File: test/integration/update_int_test.go
Comment (mdrohmann) [inline test/integration/update_int_test.go:74]: nit: I suppose that is a leftover from debugging

**Review by mdrohmann — APPROVED** [review]
Looks great. 

And https://github.com/ActiveState/cli/runs/2459251679#step:15:196 is interesting.  Is it possible though that for some users, we'll need to actually create the /Users/<user>/Application directory.  Maybe, because we are the first application in that directory?

 Well, as this happened on CI, we  certainly want to have an install option to skip the Tray installation, as a user will otherwise run into the same problem when they want to use the State Tool on CI...

File: activestate.macos.yaml
Comment (Naatan) [inline activestate.macos.yaml:15]: This seems almost identical to the non-macos version. Can we instead use a constant to tweak the flags that are different per OS?

File: activestate.macos.yaml
Comment (Naatan) [inline activestate.macos.yaml:1]: This file has a lot of stuff in it for very little change compared to the runners they're overriding. Please try to dial this back to the point that it only contains essentials.

File: internal/installation/paths.go
Comment (Naatan) [inline internal/installation/paths.go:21]: Can we name this `LauncherInstallPath()`? `System` confuses me.

File: activestate.macos.yaml
Comment (Naatan) [inline activestate.macos.yaml:26]: Isn't the default systemAppDir already correct? I don't think this runner needs anything macOS specific.

File: activestate.macos.yaml
Comment (Naatan) [inline activestate.macos.yaml:31]: @mdrohmann this is making me increasingly uncomfortable. Can you please file a followup for cleaning up the testing of updates? This is way too much mocking.

File: cmd/state-installer/internal/installer/installation_darwin.go
Comment (Naatan) [inline cmd/state-installer/internal/installer/installation_darwin.go:20]: This would produce an error if doesn't exist, I think?

File: cmd/state-installer/internal/installer/installation_test.go
Comment (Naatan) [inline cmd/state-installer/internal/installer/installation_test.go:31]: You could use this in the definitions above, then drop this whole block:

http://github.com/ActiveState/cli/blob/dd597af58d16cdac0f3cf0694338598543ec15eb/internal/exeutils/exeutils_windows.go#L3-L3

**Review by mdrohmann — APPROVED** [review]

File: scripts/update-generator/main.go
Comment (Naatan) [inline scripts/update-generator/main.go:112]: I think you can drop this and use fileutils instead:

http://github.com/ActiveState/cli/blob/dd597af58d16cdac0f3cf0694338598543ec15eb/internal/fileutils/fileutils.go#L537-L537

File: scripts/update-generator/main.go
Comment (Naatan) [inline scripts/update-generator/main.go:158]: use `ioutil.TempDir()`

**Review by Naatan — CHANGES_REQUESTED** [review]

File: activestate.macos.yaml
Comment (mdrohmann) [reply activestate.macos.yaml:31]: https://www.pivotaltracker.com/story/show/177957411

IMO this is still the best way to test the new update mechanism reliably, as it doesn't rely on "new" updates being available on S3 yet, which they are not (for the most part).  So, until the release channel has some of those updates, I would like to keep them.

I would also like to point out two advantages of these tests:
1. They test that `./scripts/update-generator` generates an installable update.  So, without these tests, we may be uploading some broken update and will only find out - after the fact.  That, of course, may not be so bad, as we should catch it on the beta-channel.
2. They make the tests much faster, as we do not have to update over the internet.

Comment (Naatan) [reply activestate.macos.yaml:31]: I'm fine with us still doing some mocking. But if I broke these tests and I had to fix this looking at this code would make me scream.. This looks daunting. I just want to get it in a place where it doesn't make me do that :p

File: cmd/state-installer/internal/installer/installation_darwin.go
Comment (MDrakos) [reply cmd/state-installer/internal/installer/installation_darwin.go:20]: No, `os.RemoveAll` will return nil if the path doesn't exist.

File: scripts/update-generator/main.go
Comment (MDrakos) [reply scripts/update-generator/main.go]: It's probably better to use that function. And in trying to use it I discovered that it would error out if there was a directory in the source path. However a lot of the logic of this function will remain as it's just creating the target path.

File: scripts/update-generator/main.go
Comment (MDrakos) [reply scripts/update-generator/main.go:141]: The installer expects a directory named `system` in its archive. I don't think that `ioutil.TempDir()` can do that as it appends a random string to the end of the filename

File: activestate.macos.yaml
Comment (MDrakos) [reply activestate.macos.yaml]: I'm not following what the default systemAppDir is that you're referring to. The value being passed in here is the one copied from the `installers` directory to the build dir

File: scripts/update-generator/main.go
Comment (Naatan) [reply scripts/update-generator/main.go:141]: you can still use `ioutil.TempDir()` for the value of `tempDir`. My worry with the current approach is that the value is constant, so you could run into conflicts.

Comment (MDrakos) [reply scripts/update-generator/main.go:141]: Fair enough, I doubled checked and `tempDir` is using `ioutil.TempDir()` https://github.com/ActiveState/cli/blob/master/scripts/update-generator/main.go#L110

File: internal/fileutils/fileutils.go
Comment (Naatan) [inline internal/fileutils/fileutils.go:590]: This looks dangerous. The `CopyAndRenameFiles` already enumerates all the files it needs to copy, and you're calling `CopyFiles` here, essentially turning all subdirs into a `CopyFiles` instruction, but while also still processing them through `CopyAndRenameFiles`.

What are you trying to solve here?

Comment (MDrakos) [reply internal/fileutils/fileutils.go:590]: When I started to use `CopyAndRenameFiles` in the update generator, this change was necessary as the function currently does not handle directories in the `fromPath`. I can revert these changes and file a follow-up story as I'm no longer using `CopyAndRenameFiles` in the update generator. It is only being used in the installer and specifically for copying the contents of the `bin` directory.

Comment (Naatan) [reply internal/fileutils/fileutils.go:604]: Yes, please revert and file a story. I'm not fully grokking what's happening here, it's looking more complicated than it probably needs to be. Either way not point holding up this PR over if you don't need it.

**Review by Naatan — CHANGES_REQUESTED** [review]
Looks good, but as noted in the other comment let's revert the fileutils changes not required for this PR.

File: internal/fileutils/fileutils.go
Comment (MDrakos) [reply internal/fileutils/fileutils.go]: Story filed here: https://www.pivotaltracker.com/story/show/178026455

**Review by Naatan — APPROVED** [review]

File: scripts/update-generator/main.go
Comment (mdrohmann) [inline scripts/update-generator/main.go:25]:
```suggestion
var outputDirFlag, platformFlag, branchFlag, versionFlag *string
```

File: scripts/update-generator/main.go
Comment (mdrohmann) [inline scripts/update-generator/main.go:126]: I think the `installerPath` is not used anymore, so it can be removed here, and also in the `createUpdate` function.

**Review by mdrohmann — DISMISSED** [review]
Looks good.  Just found two things to clean up...

**Review by mdrohmann — DISMISSED** [review]

**Review by mdrohmann — APPROVED** [review]

**Review by Naatan — APPROVED** [review]

**Pull request merged**
